pkg/logutil/zap_log_test.go

```diff
@@ -0,0 +1,229 @@
+	c.Assert(err, IsNil)
+	lg := zap.New(zapcore.NewCore(newZapTextEncoder(cfg), writer, level), opt...)
+	return verifyLogger{
+		Logger: lg,
+		w:      writer,
+	}
+}
+
+type username string
+
+func (n username) MarshalLogObject(enc zapcore.ObjectEncoder) error {
+	enc.AddString("username", string(n))
+	return nil
+}
+
+func (t *testZapLogSuite) TestLog(c *C) {
+	conf := &LogConfig{Level: "debug", File: FileLogConfig{}, DisableTimestamp: true}
+	lg := newZapTestLogger(conf, c)
+	sugar := lg.Sugar()
+	defer sugar.Sync()
+	sugar.Infow("failed to fetch URL",
+		"url", "http://example.com",
+		"attempt", 3,
+		"backoff", time.Second,
+	)
+	sugar.Infof(`failed to "fetch" [URL]: %s`, "http://example.com")
+	sugar.Debugw("Slow query",
+		"sql", `SELECT * FROM TABLE
+	WHERE ID="abc"`,
+		"duration", 1300*time.Millisecond,
+		"process keys", 1500,
+	)
+	sugar.Info("Welcome")
+	sugar.Info("Welcome TiDB")
+	sugar.Info("欢迎")
+	sugar.Info("欢迎来到 TiDB")
+	sugar.Warnw("Type",
+		"Counter", math.NaN(),
+		"Score", math.Inf(1),
+	)
+	lg.Info("Testing typs",
+		zap.String("filed1", "noquote"),
+		zap.String("filed2", "in quote"),
+		zap.Strings("urls", []string{"http://mock1.com:2347", "http://mock2.com:2432"}),
+		zap.Strings("urls-peer", []string{"t1", "t2 fine"}),
+		zap.Uint64s("store ids", []uint64{1, 4, 5}),
+		zap.Object("object", username("user1")),
+		zap.Object("object2", username("user 2")),
+		zap.Binary("binary", []byte("ab123")),
+		zap.Bool("is processed", true),
+		zap.ByteString("bytestring", []byte("noquote")),
+		zap.ByteString("bytestring", []byte("in quote")),
+		zap.Int8("int8", int8(1)),
+		zap.Uintptr("ptr", 0xa),
+		zap.Reflect("reflect", []int{1, 2}),
+		zap.Stringer("stringer", net.ParseIP("127.0.0.1")),
+		zap.Bools("array bools", []bool{true}),
+		zap.Bools("array bools", []bool{true, true, false}),
+		zap.Complex128("complex128", 1+2i),
+		zap.Strings("test", []string{
+			"💖",
+			"�",
+			"☺☻☹",
+			"日a本b語ç日ð本Ê語þ日¥本¼語i日©",
+			"日a本b語ç日ð本Ê語þ日¥本¼語i日©日a本b語ç日ð本Ê語þ日¥本¼語i日©日a本b語ç日ð本Ê語þ日¥本¼語i日©",
+			"\x80\x80\x80\x80",
+			"<car><mirror>XML</mirror></car>",
+		}),
+	)
+
+	lg.AssertMessage(
+		`[INFO] [zap_log_test.go:93] ["failed to fetch URL"] [url=http://example.com] [attempt=3] [backoff=1s]`,
+		`[INFO] [zap_log_test.go:98] ["failed to \"fetch\" [URL]: http://example.com"]`,
+		`[DEBUG] [zap_log_test.go:99] ["Slow query"] [sql="SELECT * FROM TABLE\n\tWHERE ID=\"abc\""] [duration=1.3s] ["process keys"=1500]`,
+		`[INFO] [zap_log_test.go:105] [Welcome]`,
+		`[INFO] [zap_log_test.go:106] ["Welcome TiDB"]`,
+		`[INFO] [zap_log_test.go:107] [欢迎]`,
+		`[INFO] [zap_log_test.go:108] ["欢迎来到 TiDB"]`,
+		`[WARN] [zap_log_test.go:109] [Type] [Counter=NaN] [Score=+Inf]`,
+		`[INFO] [zap_log_test.go:113] ["Testing typs"] [filed1=noquote] `+
+			`[filed2="in quote"] [urls="[http://mock1.com:2347,http://mock2.com:2432]"] `+
+			`[urls-peer="[t1,"t2 fine"]"] ["store ids"="[1,4,5]"] [object={username=user1}] `+
```

Comment:
This encoding `[urls-peer="[t1,"t2 fine"]"]` is just incorrect as well. It should be `[urls-peer="[t1,\"t2 fine\"]"]`

Reply:
This encoding `[object={username=user1}]` is also incorrect. It should be `[object="{username=user1}"]` since there is a "=" inside the value.